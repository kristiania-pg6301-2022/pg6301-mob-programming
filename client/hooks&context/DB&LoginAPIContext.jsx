import React from "react";

export const DBLoginAPIContext = React.createContext({
  async fetchLogin() {
    return await fetchJSON("/api/login");
  },
  async registerLogin(provider, login) {
    return await postJSON(`/api/login/${provider}`, login);
  },
});

export async function fetchJSON(url) {
  const res = await fetch(url);
  if (!res.ok) {
    throw new Error(`Failed to load ${res.status}: ${res.statusText}`);
  }
  return await res.json();
}
export async function postJSON(url, object) {
  const res = await fetch(url, {
    method: "POST",
    headers: {
      "content-type": "application/json",
    },
    body: JSON.stringify(object),
  });
  if (!res.ok) {
    throw new Error(`Failed to post ${res.status}: ${res.statusText}`);
  }
}
