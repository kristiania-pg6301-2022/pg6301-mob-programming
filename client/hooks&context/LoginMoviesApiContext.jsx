import React from "react";

export const LoginMoviesApiContext = React.createContext({
  async fetchLogin() {
    return await fetchJSON("/api/login");
  },
  async listMovies(query) {
    return await fetchJSON("/api/movies?" + new URLSearchParams(query));
  },
  async createMovie(movie) {
    return await postJSON("/api/movies", movie);
  },
  async registerLogin(provider, login) {
    return await postJSON(`/api/login/${provider}`, login);
  },
  async endSession() {
    const res = await fetch("/api/login", { method: "DELETE" });
    if (!res.ok) {
      throw new Error(`Failed to post ${res.status}: ${res.statusText}`);
    }
  },
  userinfo: undefined,
});

export async function fetchJSON(url) {
  const res = await fetch(url);
  if (!res.ok) {
    throw new Error(`Failed to load ${res.status}: ${res.statusText}`);
  }
  return await res.json();
}
export async function postJSON(url, object) {
  const res = await fetch(url, {
    method: "POST",
    headers: {
      "content-type": "application/json",
    },
    body: JSON.stringify(object),
  });
  if (!res.ok) {
    throw new Error(`Failed to post ${res.status}: ${res.statusText}`);
  }
}
