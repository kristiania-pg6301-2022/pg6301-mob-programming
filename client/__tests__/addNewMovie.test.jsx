import { act, Simulate } from "react-dom/test-utils";
import { MemoryRouter } from "react-router-dom";
import AddNewMovie from "../pages/addNewMovie";
import ReactDOM from "react-dom";
import React from "react";
import { LoginMoviesApiContext } from "../hooks&context/LoginMoviesApiContext";

describe("add movie", () => {
  it("should show from", async () => {
    const element = document.createElement("div");
    await act(async () =>
      ReactDOM.render(
        <MemoryRouter>
          <AddNewMovie />
        </MemoryRouter>,
        element
      )
    );
    expect(element.innerHTML).toMatchSnapshot();

    const inputLabels = Array.from(
      element.querySelectorAll("form label strong")
    ).map((label) => label.innerHTML);
    expect(inputLabels).toEqual(["Title: ", "Year:", "Country:", "Plot:"]);
  });

  it("should submit form", async () => {
    const createMovie = jest.fn();

    const element = document.createElement("div");
    await act(async () =>
      ReactDOM.render(
        <MemoryRouter>
          <LoginMoviesApiContext.Provider value={{ createMovie }}>
            <AddNewMovie />
          </LoginMoviesApiContext.Provider>
        </MemoryRouter>,
        element
      )
    );
    Simulate.change(element.querySelector("form div:nth-of-type(1) input"), {
      target: { value: "Movie Title" },
    });
    Simulate.change(element.querySelector("form div:nth-of-type(2) input"), {
      target: { value: "2022" },
    });
    Simulate.submit(element.querySelector("form"));

    expect(createMovie).toBeCalledWith({
      title: "Movie Title",
      year: 2022,
      plot: "",
      country: "",
    });
  });
});
