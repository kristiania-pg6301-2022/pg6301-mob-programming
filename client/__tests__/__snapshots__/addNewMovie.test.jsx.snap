// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add movie should show from 1`] = `"<form><h1>Add a new movie</h1><div><label><strong>Title: </strong><input value=\\"\\"></label></div><div><label><strong>Year:</strong><input value=\\"\\"></label></div><div><label><strong>Country:</strong><input value=\\"\\"></label></div><div><label><strong>Plot:</strong><input value=\\"\\"></label></div><button>Save</button></form>"`;
